service/matching/tasklist/task_completer.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	taskCompleterImpl struct {
+		domainCache     cache.DomainCache
+		taskListID      *Identifier
+		clusterMetadata cluster.Metadata
+		historyService  history.Client
+		scope           metrics.Scope
+		logger          log.Logger
+		throttleRetry   *backoff.ThrottleRetry
+	}
+)
+
+// DomainIsActiveInThisClusterError type
+type DomainIsActiveInThisClusterError struct {
+	Message string
+}
+
+// Implement the Error method for DomainIsActiveInThisClusterError
+func (e *DomainIsActiveInThisClusterError) Error() string {
+	return e.Message
+}
+
+var (
+	errWorkflowExecutionInfoIsNil      = errors.New("workflow execution info is nil")
+	errTaskTypeNotSupported            = errors.New("task type not supported")
+	errTaskNotStarted                  = errors.New("task not started")
+	errDomainIsActive                  = &DomainIsActiveInThisClusterError{Message: "domain is active"}
+	historyServiceOperationRetryPolicy = common.CreateTaskCompleterRetryPolicy()
+)
+
+func newTaskCompleter(tlMgr *taskListManagerImpl, retryPolicy backoff.RetryPolicy) TaskCompleter {
+	return &taskCompleterImpl{
+		domainCache:     tlMgr.domainCache,
+		historyService:  tlMgr.historyService,
+		taskListID:      tlMgr.taskListID,
+		clusterMetadata: tlMgr.clusterMetadata,
+		scope:           tlMgr.scope,
+		logger:          tlMgr.logger,
+		throttleRetry: backoff.NewThrottleRetry(
+			backoff.WithRetryPolicy(retryPolicy),
+			backoff.WithRetryableError(isRetryableError),
+		),
+	}
+}
+
+func (tc *taskCompleterImpl) CompleteTaskIfStarted(ctx context.Context, task *InternalTask) error {
+	op := func() (err error) {
+		domainEntry, err := tc.domainCache.GetDomainByID(task.Event.TaskInfo.DomainID)
+		if err != nil {
+			return fmt.Errorf("unable to fetch domain from cache: %w", err)
+		}
+
+		if _, err = domainEntry.IsActiveIn(tc.clusterMetadata.GetCurrentClusterName()); err == nil {
+			return errDomainIsActive
+		}
+
+		req := &types.HistoryDescribeWorkflowExecutionRequest{
+			DomainUUID: task.Event.TaskInfo.DomainID,
+			Request: &types.DescribeWorkflowExecutionRequest{
+				Domain: task.domainName,
+				Execution: &types.WorkflowExecution{
+					WorkflowID: task.Event.WorkflowID,
+					RunID:      task.Event.RunID,
+				},
+			},
+		}
+
+		workflowExecutionResponse, err := tc.historyService.DescribeWorkflowExecution(ctx, req)
+
+		if errors.As(err, new(*types.EntityNotExistsError)) {
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("unable to fetch workflow execution from the history service: %w", err)
+		}
+
+		if workflowExecutionResponse.WorkflowExecutionInfo == nil {
+			return errWorkflowExecutionInfoIsNil
+		}
+
+		isTaskStarted, err := tc.isTaskStarted(task, workflowExecutionResponse)
+		if err != nil {
+			return fmt.Errorf("unable to determine if task has started: %w", err)
+		}
+
+		if isTaskStarted {
+			return nil
+		}
+
+		tc.scope.
+			Tagged(metrics.DomainTag(task.domainName)).
```

Comment:
giga-Nit: I think we might get these tags from the current scope since we're using the tasklist manager's scope. If not I think it would make sense to construct a new scope so that we have a consistent set of dimensions for all these metrics if we forget to include them.